lib/spack/spack/spec.py

```diff
@@ -4845,26 +4845,35 @@ def reconstruct_virtuals_on_edges(spec):
         spec: spec on which we want to reconstruct virtuals
     """
     # Collect all possible virtuals
-    possible_virtuals = set()
+    possible_virtuals = collections.defaultdict(set)
     for node in spec.traverse():
+        key = node.dag_hash()
         try:
-            possible_virtuals.update(
-                {x for x in node.package.dependencies if spack.repo.PATH.is_virtual(x)}
-            )
+            for name, when_deps in node.package.dependencies_by_name(when=True).items():
+                if not spack.repo.PATH.is_virtual(name):
+                    continue
+
+                if any(node.satisfies(x) for x in when_deps):
+                    possible_virtuals[key].add(name)
         except Exception as e:
-            warnings.warn(f"cannot reconstruct virtual dependencies on package {node.name}: {e}")
+            warnings.warn(f"cannot reconstruct virtual dependencies on {node.name}: {e}")
```

Comment:
I think you wanna just wrap `node.package` in try-catch? I don't see why the whole block warrants it

Reply:
We don't want that failing to assign a virtual raises, so this was out of caution. I just wrapped `node.package` now.